RISC-V vector crypto support

cmuellner opened the pull request:
This PR adds support for accelerated cryptographic routines using the [RISC-V vector crypto extensions](https://github.com/riscv/riscv-crypto/releases/tag/v20230620).
The following optimized routines are provided in the PR:
* GCM: Zvbb and Zvbc (carry-less multiplication)
* GCM: Zvkg (GCM/GMAC instructions - optional with Zvbb)
* AES: Zvkns
* SHA-256: Zvbb and Zvknha
* SHA-512: Zvbb and Zvknhb
* SM4: Zvbb and Zvksed
* SM3: Zvbb and Zvksh

Extension support for the RISC-V vector crypto extensions has been merged upstream to [Binutils](https://sourceware.org/git/?p=binutils-gdb.git;a=commit;h=704b30cbb2beeac1e90c4ee6f1f5b9f1dc2e5ee4) and [GCC](https://gcc.gnu.org/git/?p=gcc.git;a=commit;h=7c521f6751e47d9dfaca2c8288c65f94e2f54d3d).

A QEMU patchset on the [QEMU mailing list](https://lists.gnu.org/archive/html/qemu-devel/2023-07/msg00136.html) can be used to test this PR. This PR has been tested with [this patchset with two patches on top](https://github.com/cmuellner/qemu/tree/riscv-zvk-v7).

cmuellner commented:
Updated the PR:
* Added AES implementation
* moved common Perl-Asm code into a Perl module (like other architectures have one as well)
* make Perl-Asm running with `use strict`
* rebased on `upstream/master`

cmuellner commented:
Updated the PR:

* Added SM3/Zvksh support
* Rebased on upstream/master
* A few small cleanups

All individual SM4 and SM3 test cases are passing.
However, we have observed that one combined test case is failing: test_pkcs8.
This issue is currently under investigation.

nibrunieAtSi5 reviewed (commented):
Beware that `Zvkns` has been renamed to `Zvkned` in RISC-V vector crypto specification.

egouriou-rivos commented on include/crypto/riscv_arch.h:
```diff
@@ -61,4 +61,7 @@ static const size_t kRISCVNumCaps =
 #define RISCV_HAS_ZBKB_AND_ZKND_AND_ZKNE() (RISCV_HAS_ZBKB() && RISCV_HAS_ZKND() && RISCV_HAS_ZKNE())
 #define RISCV_HAS_ZKND_AND_ZKNE() (RISCV_HAS_ZKND() && RISCV_HAS_ZKNE())
 
+/* Get the size of a vector register in bits (VLEN). */
```
Minor: is it legal to call it when RISCV_HAS_V() is false? Do we want to guarantee that the result is 0 then?

cmuellner replied:
Is not documented accordingly.

egouriou-rivos commented on include/crypto/riscv_arch.def:
```diff
@@ -32,6 +32,7 @@ RISCV_DEFINE_CAP(ZKSED, 0, 10)
 RISCV_DEFINE_CAP(ZKSH, 0, 11)
 RISCV_DEFINE_CAP(ZKR, 0, 12)
 RISCV_DEFINE_CAP(ZKT, 0, 13)
+RISCV_DEFINE_CAP(V, 0, 14)
```
I do not see any code here to set this capability based on a runtime query.

If this is intended, consider clarifying the commit description which can be interpreted to mean that "riscv_vlen()" is valid after this commit is included.
> If this extension is found at run-time, then we read out VLEN as described above and cache it.

cmuellner replied:
The code is already in OpenSSL (`crypto/riscvcap.c`).
The patch series is just reusing the existing mechanism.

cmuellner replied:
Regarding `riscv_vlen()`: this is addressed by specifying the behaviour of that function if no vector extension is found (i.e. returning 0).

egouriou-rivos reviewed (commented):
Initial comments after checking the first 7 commit, before going in the meat of the Zvk code. I only glanced at the scalar crypto changes to GCM, will come back to it.

Only minor comments/suggestions so far. Nice cleanup overall.

JerryShih commented on crypto/modes/asm/ghash-riscv64-zvkb.pl:
```diff
@@ -0,0 +1,349 @@
+# void gcm_init_rv64i_zvkg(u128 Htable[16], const u64 H[2]);
+#
+# input:	H: 128-bit H - secret parameter E(K, 0^128)
+# output:	Htable: Preprocessed key data for gcm_gmult_rv64i_zvkb and
```
Could you describe the content of Htable in `gcm_init_rv64i_zvkb`?
Does this `gcm_init_rv64i_zvkb` do the byte reversion as `gcm_init_rv64i_zvkg`?

JerryShih commented on crypto/modes/asm/ghash-riscv64-zvkg.pl:
```diff
@@ -0,0 +1,161 @@
+    @{[vle32_v $vinp, $inp]}
+    add $inp, $inp, 16
+    add $len, $len, -16
+    @{[vghsh_vv $vXi, $vinp, $vH]}
```
Should we need to swap the `$vinp` and `$vH` position?
```
vghsh.vv vd, vs2, vs1
=>
vd = (vd^vs1)*vs2
```

----
spec:
![image](https://user-images.githubusercontent.com/5842681/222955189-7402652f-f6a5-4cdf-8dfa-91e6d8173feb.png)

cmuellner replied:
You are right here.
I proposed a modification of the instruction (doing XOR before multiplication instead of otherwise) to the TG, which was accepted. However, the specification has adopted my proposal with vs1 and vs2 in a different order, which I overlooked.

So our internal QEMU implementation of Zvk* had this different order as well, which is why testing did not reveal this as an issue.

JerryShih commented on crypto/sha/asm/sha512-riscv64-zvknhb.pl:
```diff
@@ -0,0 +1,347 @@
+.globl sha512_block_data_order_zvknhb
+.type sha512_block_data_order_zvknhb,\@function
+sha512_block_data_order_zvknhb:
+    @{[vsetivli__x0_4_e64_m1_ta_ma]}
```
Will we consider the VLEN=128 platform in the future? We only use 11 vector registers here. We could use LMUL=2 here.

```
my ($V0, $V10, $V11, $V12, $V13, $V14, $V15, $V16, $V17) = ("v0", "v10", "v11", "v12", "v13", "v14","v15", "v16", "v17");
my ($V26, $V27) = ("v26", "v27");
```

nibrunieAtSi5 replied:
That's a good point. The vector crypto standard does not require VLEN >= 256 for SHA-512 so this code should be written to be compatible with VLEN=128 if possible (vl=4, SEW=64-bit, LMUL=2) to be compatible with any VLEN >= 128. As it seems the V extension requires only VLEN >= 128-bit, https://github.com/riscv/riscv-v-spec/blob/master/v-spec.adoc#183-v-vector-extension-for-application-processors

cmuellner replied:
Support for LMUL=2 is fine, but we need to add that **additionally** to this implementation. HW that implements bigger vector registers should not suffer from overhead introduced to support VLEN=128 as well.

JerryShih commented on crypto/sm3/asm/sm3-riscv64-zvksh.pl:
```diff
@@ -0,0 +1,195 @@
+.globl ossl_hwsm3_block_data_order_zvksh
+.type ossl_hwsm3_block_data_order_zvksh,\@function
+ossl_hwsm3_block_data_order_zvksh:
+    @{[vsetivli__x0_8_e32_m1_ta_ma]}
```
Should we consider the VLEN=128 platform?

cmuellner replied:
Same as above: HW with VLEN=256 must not see any penalties.

JerryShih commented on crypto/modes/asm/ghash-riscv64-zvkg.pl:
```diff
@@ -0,0 +1,161 @@
+# output: Htable: Copy of secret parameter (in normalized byte order)
+#
+# All callers of this function revert the byte-order unconditionally
+# on little-endian machines. So we need to revert the byte-order back.
```
I am not familiar with GCM mode in openssl.
Do you know the reason that we always do the byte reversion for H?

```
void CRYPTO_gcm128_init(GCM128_CONTEXT *ctx, void *key, block128_f block)
{
    DECLARE_IS_ENDIAN;

    memset(ctx, 0, sizeof(*ctx));
    ctx->block = block;
    ctx->key = key;

    (*block) (ctx->H.c, ctx->H.c, key);

    if (IS_LITTLE_ENDIAN) {
        /* H is stored in host byte order */
#ifdef BSWAP8
        ctx->H.u[0] = BSWAP8(ctx->H.u[0]);
        ctx->H.u[1] = BSWAP8(ctx->H.u[1]);
#else
        u8 *p = ctx->H.c;
        u64 hi, lo;
        hi = (u64)GETU32(p) << 32 | GETU32(p + 4);
        lo = (u64)GETU32(p + 8) << 32 | GETU32(p + 12);
        ctx->H.u[0] = hi;
        ctx->H.u[1] = lo;
#endif
    }

    gcm_get_funcs(&ctx->funcs);
    ctx->funcs.ginit(ctx->Htable, ctx->H.u);
}
```

cmuellner replied:
I was asking myself this question as well two months ago.
I concluded that it makes sense for non-optimized ASM code, but OpenSSL's requirement is wrong here.

I did not clean this up, because I considered it out of scope for this patchset (and potentially a discussion point that prevents this PR from getting merged).

I fully support any attempts to clean this up. My proposed approach would be: Make this not depend on `IS_LITTLE_ENDIAN`, but `GCM_INIT_SWAP_H` which defaults to `IS_LITTLE_ENDIAN` but can be defined differently by platform code

egouriou-rivos commented on crypto/aes/asm/aes-riscv64-zvkned.pl:
```diff
@@ -0,0 +1,500 @@
+    @{[vaeskf1_vi $v19, $v18, 9]}   # v19 <- rk10 (w[36,39])
+    @{[vaeskf1_vi $v20, $v19, 10]}  # v20 <- rk11 (w[40,43])
+
+    # Store the round keys
```
An in-order processor would probably benefit from interleaving the key schedule operations and the stores.
The current code has the benefit of being clear.

cmuellner replied:
I was thinking about this as well when writing the code.

However, my current view on this is: make the code clear and obvious (i.e. enable extensions) and do the tweaking later guided by performance measurements on real HW.

egouriou-rivos commented on crypto/aes/asm/aes-riscv64-zvkned.pl:
```diff
@@ -0,0 +1,500 @@
+
+    @{[vmv_v_v $v12, $v10]}
+    @{[vaeskf2_vi $v12, $v11, 1]}
+    @{[vmv_v_v $v13, $v11]}
```
Pretty much all vmv_v_v can be moved up to hide their latency from the consuming vaeskf2. Not an issue on OoO vector implementations.

cmuellner replied:
Same as above.

JerryShih commented on crypto/modes/asm/ghash-riscv64-zvkb.pl:
```diff
@@ -0,0 +1,349 @@
+    @{[vmerge_vim $V3, $V3, 3]}      # vmerge.vim v3, v3, 3, v0
+    @{[vmv_v_v $V0, $V3]}            # vmv.v.v v0, v3
+
+    @{[vxor_vv_v0t $V1, $V1, $V2]}   # vxor.vv v1, v1, v2, v0.t
```
I don't know why we need to use a mask with 3 value. The 3 value means that we do the operation on element index 0 and element index 1.
The vl is 2, and we need to do the xor for all elements between v1 and v2.
We could just do the `vxor` directly.
```
vxor.vv v1, v1, v2
```

JerryShih replied:
What's the `polymod` used for?
Does it use for the remainder of α^128?
The remainder should be `1 + α + α^2 + α^7`.
GCM uses the polynomial `f = 1 + α + α^2 + α^7 + α^128` for mod.

```
.p2align 4
Lpolymod:
        .dword 0x0000000000000001
        .dword 0xc200000000000000
.size Lpolymod,.-Lpolymod
```

cmuellner replied:
`polymod` is used for reducing the resulting value after multiplication.
See also https://github.com/rivosinc/riscv-code-samples/blob/main/zvk-vector-crypto/zvkb.s#L754

cmuellner commented on crypto/perlasm/riscv.pm:
```diff
@@ -9,20 +9,41 @@
+}
+
 my @regs = map("x$_",(0..31));
+my @regaliases = ('zero','ra','sp','gp','tp','t0','t1','t2','s0','s1',
```
I think people that don't understand the mapping are probably not aware of the psABI document.
Therefore I added the following comment, which provides a proper search term to lead them to the required information: `Mapping from the RISC-V psABI ABI mnemonic names to the register number.`

cmuellner commented on crypto/modes/asm/ghash-riscv64-zvkb.pl:
```diff
@@ -0,0 +1,349 @@
+___
+
+################################################################################
+# void gcm_init_rv64i_zvkg(u128 Htable[16], const u64 H[2]);
```
Fixed.

cmuellner commented on crypto/modes/asm/ghash-riscv64-zvkb.pl:
```diff
@@ -0,0 +1,349 @@
+    li $TMP1, 63
+    la $TMP2, Lpolymod
+
+    @{[vsetivli__x0_2_e64_m1_ta_ma]} # vsetivli x0, 2, e64, m1, ta, ma
```
Zve32 and Zve64 are not supported by this PR in the sense that no optimized Zvk* routines will be executed.

The top of all files lists the requirements. In the case of `ghash-riscv64-zvkb.pl`:
```
# - RV64I
# - RISC-V vector ('V') with VLEN >= 128
# - Vector Bit-manipulation used in Cryptography ('Zvkb')
```

And runtime checks ensure that VLEN is big enough (`riscv_vlen() >= 128`).

So the PR requires reasonable and documented constraints for VLEN, which are tested at runtime.

cmuellner commented on crypto/modes/asm/ghash-riscv64-zvkb.pl:
```diff
@@ -0,0 +1,349 @@
+
+    # Use the fact that the polynomial degree is no more than 128,
+    # i.e. only the LSB of the upper half could be set.
+    # Thanks to we don't need to do the full reduction here.
```
Fixed.

cmuellner commented on crypto/sm3/asm/sm3-riscv64-zvksh.pl:
```diff
@@ -0,0 +1,195 @@
+    # As vsm3c consumes only w0, w1, w4, w5 we need to slide the input
+    # 2 elements down so we process elements w2, w3, w6, w7
+    # This will be repeated for each odd round.
+    @{[vslidedown_vi $V2, $V3, 2]} # v2 := {w2, ..., w7, 0, 0}
```
Agreed. Using `0` is misleading. Will use `X` instead for `unknown`.

However, the code is correct, as it does not rely on the elements to be `0`, but treats them as `X` and does not process these elements.

cmuellner replied:
This was fixed in a previous version.

cmuellner commented on crypto/sm3/asm/sm3-riscv64-zvksh.pl:
```diff
@@ -0,0 +1,195 @@
+    @{[vle32_v $V3, $INPUT]} # v3 := {w0, ..., w7}
+    add $INPUT, $INPUT, 32
+
+    @{[vle32_v $V4, $INPUT]} # v4 := {w8, ..., w15}
```
My view of this is that we have an array or a  list of elements.
As far as I can tell, the notation used is consistent and correct.

If you want to, I can change the document, but I would prefer not to (not because I have a preference, but because I want to avoid introducing a mistake when converting).

cmuellner replied:
Fixed.

cmuellner commented on crypto/sm3/asm/sm3-riscv64-zvksh.pl:
```diff
@@ -0,0 +1,195 @@
+
+    # Load initial state of hash context (c->A-H).
+    @{[vle32_v $V0, $CTX]}
+    @{[vrev8_v $V0, $V0]}
```
Yes, we don't control the endianess of the state in the Zvksh code.
And the state needs to remain compatible with the platform-independent code.

cmuellner commented on crypto/sha/asm/sha512-riscv64-zvknhb.pl:
```diff
@@ -0,0 +1,347 @@
+    addi $H, $H, -32
+
+    @{[vmv_v_v $V27, $V16]} # {d,c,b,a}
+    @{[vslidedown_vi $V26, $V16, 2]} # {b,a,0,0}
```
Fixed.

JerryShih commented on crypto/modes/asm/ghash-riscv64-zvkb.pl:
```diff
@@ -0,0 +1,349 @@
+    # B = b1b0 (v5)
+    # C = c1c0 (256 bit)
+    # c1 = a1b1 + (a0b1)h + (a1b0)h
+    # c0 = a0b0 + (a0b1)l + (a1b0)h
```
I'm confused for these comment.
I think it should be:
```
    # c1 = a1b1 + (a0b1)h + (a1b0)h
    # c0 = a0b0 + (a0b1)l << 64 + (a1b0)l << 64
```

cmuellner commented:
@egouriou-rivos  and @JerryShih: thank you both for the review!

I tried to address all comments except for the GCM/Zvkb implementation, which strictly follows the code in https://github.com/rivosinc/riscv-code-samples.

I've adjusted the code according to the feedback, rebased and retested it.

As mentioned in the PR description, this PR sits on a couple of clean-up commits, that could already land in OpenSSL. These commits are part of the following PR: https://github.com/openssl/openssl/pull/20078. It would be great if you could review that so that we can land at least some commits of this series.

phoebesv commented:
I have cross-compiled using riscv-unknown-linux-gnu-gcc toolchain as indicated by the following configuration:
`Configure linux64-riscv64 --cross-compile-prefix=riscv64-unknown-linux-gnu- --prefix=/path/to/openssl/build -mabi=lp64d --strict-warnings enable-fips`
and use openssl speed to test AES GCM for this PR by the following command:
`OPENSSL_riscvcap=OPENSSL_riscvcap_v_zvkg  qemu-riscv64 -cpu rv64,g=false,f=true,d=true,v=true,vlen=128,elen=64,vext_spec=v1.0,x-zvkb=true,x-zvkg=true,x-zvknha=true,x-zvknhb=true,x-zvkned=true,x-zvksed=true,x-zvksh=true -L ./path/to/toolchain/sysroot/ -E LD_LIBRARY_PATH=./lib ./openssl speed -elapsed -evp aes-128-gcm -seconds 30`
An error has been encountered in openssl/apps/speed.c :778 "Error finalizing cipher loop". I am unsure whether there is a problem with the cipher GCM? Would it be necessary to make any changes to the configuration or command?

phoebesv commented:
> @phoebesv, which QEMU patchset did you use?
> 
> My test process also uses cross-compilation and executes `make test`. I will try to get `openssl speed` running as well (haven't tested this so far).

@cmuellner, this is the patchset that I utilized - https://patchew.org/QEMU/20230310091215.931644-1-lawrence.hunter@codethink.co.uk/
Upon further investigation, it has been determined that the error is unrelated to this pull request, but rather the underlying cause of the issue appears to be with the error message condition in OpenSSL, specifically located at https://github.com/openssl/openssl/commit/98283a61f5b792dde0f0e1d9a616ec6e232b57b6#diff-a7b829cf9e873b0444f494b7224698d27fbbccc1a9f87e78210e6eb9d69f8c79R777-R778." 
The correct condition should be `if (rc <= 0)`.

phoebesv commented:
@cmuellner Thank you for testing, I have noticed that the issue has been resolved in the master branch.(https://github.com/openssl/openssl/pull/20306) openssl-speed issue could be resolved after rebasing to the latest master commit.

cmuellner commented:
Updated the patchset:
* Perl asm: switch license from Apache to dual Apache+BSD, so that the Linux kernel can later adopt the files (see also #20649 why this is necessary)
* rebase on master and resolve merge issue

Retesting worked well.

There have been changes to the specification recently, which also affect this PR (e.g. vclmul* instructions moved into their own extension). These changes are not included in this update.

The Zvk specification is currently under final review by RISC-V's architectural review committee. Once this passes, the specification can go towards the freeze milestone (which means that changes are very unlikely from there on).

openssl-machine commented:
This PR is waiting for the creator to make requested changes but it has not been updated for 30 days.  If you have made changes or commented to the reviewer please make sure you re-request a review (see icon in the 'reviewers' section).

openssl-machine commented:
This PR is waiting for the creator to make requested changes but it has not been updated for 61 days.  If you have made changes or commented to the reviewer please make sure you re-request a review (see icon in the 'reviewers' section).

JerryShih commented on crypto/modes/asm/ghash-riscv64-zvkb.pl:
```diff
@@ -0,0 +1,349 @@
+    add $Xi, $Xi, 8
+    li $TMP4, -8
+
+    @{[vsetivli__x0_2_e64_m1_ta_ma]} # vsetivli x0, 2, e64, m1, ta, ma
```
Could we just use the `standard vector extension` directly instead of the opcode?
I think gcc and llvm assembler all support vector 1.0 spec.

If we try to have more riscv-optimized ciphers in openssl, it's annoying to put all combination of opcodes for `standard vector extension` in further developing.

paulidale replied:
Do _all_ versions of gcc and llvm _in use_ support these instructions?

paulidale replied:
The project supports many very old versions.  I don't think you need to do the same but it does preclude using instructions that likely aren't present everywhere.  RISC-V is new enough that this isn't a complete disaster (yet).

cmuellner replied:
I agree with @paulidale, because we simply have to support "very old compiler".
Using an old compiler should not be a reason to not use the latest version of OpenSSL.

JerryShih commented:
@cmuellner

The public review of vector crypto spec has started at Friday June 23rd for 30 days.
https://groups.google.com/a/groups.riscv.org/g/isa-dev/c/DpjkaK_1zQs/m/ubf7oSBFAgAJ?utm_medium=email&utm_source=footer

Will we start to update this pr with the last spec recently?

phoebesv commented on crypto/modes/asm/ghash-riscv64-zvkb.pl:
```diff
@@ -0,0 +1,349 @@
+
+# - RV64I
+# - RISC-V vector ('V') with VLEN >= 128
+# - Vector Bit-manipulation used in Cryptography ('Zvkb')
```
ZVKB vector bit-manipulation extension has been renamed to zvbb in the latest vector crypto frozen specification 1.0.0-rc1. https://github.com/riscv/riscv-crypto/releases/latest
The zvkb terms in the PR may need to be renamed.

cmuellner replied:
Updated.

cmuellner commented on crypto/aes/asm/aes-riscv64-zvkned.pl:
```diff
@@ -0,0 +1,500 @@
+    @{[vle32_v $v11, ($UKEY)]}
+
+    @{[vmv_v_v $v12, $v10]}
+    @{[vaeskf2_vi $v12, $v11, 1]}
```
You are right. This will be fixed.

cmuellner commented:
I've updated this PR's description as follows:
* Rewrote the feature description to follow the frozen specification
* Add link to the frozen specification
* Add links to landed RISC-V vector crypto support for Binutils and GCC
* Add links to QEMU patchset for RISC-V vector crypto support (plus a link to a remote branch in a GitHub repo)

Further, I have updated the PR to follow the frozen specification of the vector crypto extensions:
* Address extension renamings and regrouping
* Fix issue with clarified behaviour of `aeskf2`
* Rebase on `upstream/master`

JerryShih commented on include/crypto/sm4_platform.h:
```diff
@@ -38,7 +38,19 @@ static inline int vpsm4_ex_capable(void)
+#  elif defined(__riscv) && __riscv_xlen == 64
+/* RV64 support */
+#   include "riscv_arch.h"
+/* Zvkns extension (vector crypto SM4). */
```
Zvkns => Zvksed

cmuellner replied:
Fixed

phoebesv commented on crypto/modes/gcm128.c:
```diff
@@ -413,6 +413,17 @@ void gcm_ghash_rv64i_zbc(u64 Xi[2], const u128 Htable[16],
+void gcm_gmult_rv64i_zvbb_zvbc(u64 Xi[2], const u128 Htable[16]);
+void gcm_ghash_rv64i_zvbb_zvbc(u64 Xi[2], const u128 Htable[16],
+                               const u8 *inp, size_t len);
+/* Zvkg (vector crypto with vghmac.vv). */
```
vghmac has been renamed to vghsh in RISC-V vector crypto specification.

cmuellner replied:
Fixed

cmuellner commented on crypto/modes/asm/ghash-riscv64-zvkb.pl:
```diff
@@ -0,0 +1,349 @@
+    @{[vslidedown_vi $V3, $V3, 1]}   # vslidedown.vi v3, v3, 1
+
+    @{[vmv_v_i $V0, 2]}              # vmv.v.i v0, 2
+    @{[vor_vv_v0t $V1, $V1, $V4]}    # vor.vv v1, v1, v4, v0.t
```
The vslideup moves only by one element (`vslideup.vi v4, v3, 1`), but the mask constant is 2 (`vmv.v.i v0, 2`). So we need to keep the masked vor. Do I miss something?

cmuellner commented:
Updated PR:
* Address some of the unresolved comments
* Update license header to allow reuse in the Linux kernel (see #21357)

JerryShih commented:
Public review of vector crypto has ended(7/23).
Will we review/merge this pr soon?

paulidale commented:
First up, click the "Ready for review" button down the bottom.
Then two of the OpenSSL team will need to review this.
Given that none of us are acquainted with RISC-V, it would be very helpful if someone who was also reviewed the code.

cmuellner commented:
End of public review does not mean ratified. Comments have been collected and answered, some minor spec changes might be still outstanding. Next comes the final ARC review and the approval of the TSC. Anyway, I will make the PR as "ready for review" because I think the time has come for that.

FWIW, Heiko has posted the Perl-routines on the LKML (for inclusion in the Linux kernel), and they are (independently) reviewed there as well. My goal is to ensure that OpenSSL and Linux will have the same Perl code (unless there are good reasons against that).

cmuellner commented:
Ard Biesheuvel filed a PR against my GitHub repo (https://github.com/cmuellner/openssl/pull/1) to add AES-192 support using not-accelerated key expansion and accelerated encryption/decryption.

I've reviewed/tested the PR and included it into the Zvk patchset (after rebasing on latest master).

nibrunieAtSi5 commented:
> Ard Biesheuvel filed a PR against my GitHub repo ([cmuellner#1](https://github.com/cmuellner/openssl/pull/1)) to add AES-192 support using not-accelerated key expansion and accelerated encryption/decryption.
> 
> I've reviewed/tested the PR and included it into the Zvk patchset (after rebasing on latest master).

That is quite interesting. AES-192 was definitely not a target for `Zvk`.

phoebesv commented:
@cmuellner We appreciate your contribution to the RVV crypto optimization code. Recently, we tested AES-GCM implementations with RISC-V hardware capabilities enabled using `$OPENSSL_riscvcap=OPENSSL_riscvcap_v_zvbb_zvbc_zvkned` on qemu, but encountered unexpected results in the evp_test gcm tests. AES-GCM test cases can be found in the openssl/test/recipes/30-test_evp_data/evpciph_aes_common.txt. 
test_evp test binary is executed using qemu by the following command:
```
OPENSSL_riscvcap=OPENSSL_riscvcap_v_zvbb_zvbc_zvkned \
qemu-riscv64 -cpu rv64,g=false,f=true,d=true,v=true,vlen=128,elen=64,vext_spec=v1.0,x-zvbb=true,x-zvbc=true,x-zvkg=true,x-zvknha=true,x-zvknhb=true,x-zvkned=true,x-zvksed=true,x-zvksh=true,zbkb=true
./test/evp_test ./test/recipes/30-test_evp_data/evpciph_aes_common.txt
```
Would you be able to assist with these problems? In addition, I am interested in knowing how you validate openssl crypto accuracy on QEMU.

cmuellner commented:
@phoebesv, my testing approach is to test each individual patch with `make test` and the expected result of all tests passing. The QEMU tree, which was used, is https://github.com/cmuellner/qemu/tree/riscv-zvk-v7.

To give you an overview of all combinations that I test, here is the relevant excerpt of one of my test scripts:
```
../build.sh build

../build.sh test-gcm qemu-cpu-zvbb-zvbc openssl-cap-zvbb-zvbc
../build.sh test-gcm qemu-cpu-zvkg openssl-cap-zvkg
../build.sh test-gcm qemu-cpu-zvbb-zvkg openssl-cap-zvbb-zvkg
../build.sh test-aes qemu-cpu-zvkned openssl-cap-zvkned
../build.sh test-sha qemu-cpu-zvbb-zvknha openssl-cap-zvbb-zvknha
../build.sh test-sha qemu-cpu-zvbb-zvknhb openssl-cap-zvbb-zvknhb
../build.sh test-sm4 qemu-cpu-zvbb-zvksed openssl-cap-zvbb-zvksed
../build.sh test-sm3 qemu-cpu-zvbb-zvksh openssl-cap-zvbb-zvksh
../build.sh test-all qemu-cpu-zvk openssl-cap-zvk
```

I tried to reproduce your issue, but I failed:
```
$ echo $OPENSSL_riscvcap
rv64gcv_zvbb_zvbc_zvkned
$ qemu-riscv64 -cpu rv64,g=false,f=true,d=true,v=true,vlen=128,elen=64,vext_spec=v1.0,zvbb=true,zvbc=true,zvkg=true,zvknha=true,zvknhb=true,zvkned=true,zvksed=true,zvksh=true,zbkb=true  ./test/evp_test ./test/recipes/30-test_evp_data/evpciph_aes_common.txt

    # Completed 164 tests with 0 errors and 1 skipped
    ok 1 - iteration 1
ok 1 - run_file_tests
```

Can you show the error message that your are facing?
Is it an QEMU issue (e.g. illegal instruction) or a wrong result?
Also, which QEMU sources are you using? There were lots of patchsets for older spec versions and with bugs around. So make sure you have a reliable one.

phoebesv commented:
@cmuellner I think the root cause of the failure of the test is the incorrect mask policy in ghash-riscv64-zvbb-zvbc.pl. It should be set as mask-undisturbed. Tail/mask-agnostic setting should be aware, since some CPUs will force it to a default value. You can test if the tail or mask policy is incorrect by adding rvv_ma_all_1s=true and rvv_ta_all_1s=true to the qemu argument. This is helpful for identifying potential issues with tails and masks.

phoebesv commented on crypto/modes/asm/ghash-riscv64-zvbb-zvbc.pl:
```diff
@@ -0,0 +1,378 @@
+    li $TMP1, 63
+    la $TMP2, Lpolymod
+
+    @{[vsetivli__x0_2_e64_m1_ta_ma]} # vsetivli x0, 2, e64, m1, ta, ma
```
vsetivli x0, 2, e64, m1, ta, ma =>  vsetivli x0, 2, e64, m1, ta, mu

phoebesv commented on crypto/modes/asm/ghash-riscv64-zvbb-zvbc.pl:
```diff
@@ -0,0 +1,378 @@
+    add $Xi, $Xi, 8
+    li $TMP4, -8
+
+    @{[vsetivli__x0_2_e64_m1_ta_ma]} # vsetivli x0, 2, e64, m1, ta, ma
```
vsetivli x0, 2, e64, m1, ta, ma => vsetivli x0, 2, e64, m1, ta, mu

phoebesv commented on crypto/modes/asm/ghash-riscv64-zvbb-zvbc.pl:
```diff
@@ -0,0 +1,378 @@
+    add $inp, $inp, 8
+    li $M8, -8
+
+    @{[vsetivli__x0_2_e64_m1_ta_ma]} # vsetivli x0, 2, e64, m1, ta, ma
```
vsetivli x0, 2, e64, m1, ta, ma => vsetivli x0, 2, e64, m1, ta, mu

charmitro commented:
> @cmuellner We appreciate your contribution to the RVV crypto optimization code. Recently, we tested AES-GCM implementations with RISC-V hardware capabilities enabled using `$OPENSSL_riscvcap=OPENSSL_riscvcap_v_zvbb_zvbc_zvkned` on qemu, but encountered unexpected results in the evp_test gcm tests. AES-GCM test cases can be found in the `openssl/test/recipes/30-test_evp_data/evpciph_aes_common.txt`. test_evp test binary is executed using qemu by the following command:
> 
> ```
> OPENSSL_riscvcap=OPENSSL_riscvcap_v_zvbb_zvbc_zvkned \
> qemu-riscv64 -cpu rv64,g=false,f=true,d=true,v=true,vlen=128,elen=64,vext_spec=v1.0,x-zvbb=true,x-zvbc=true,x-zvkg=true,x-zvknha=true,x-zvknhb=true,x-zvkned=true,x-zvksed=true,x-zvksh=true,zbkb=true,rvv_ma_all_1s=true,rvv_ta_all_1s=true
> ./test/evp_test ./test/recipes/30-test_evp_data/evpciph_aes_common.txt
> ```
> 
> Would you be able to assist with these problems? In addition, I am interested in knowing how you validate openssl crypto accuracy on QEMU.

I can reproduce your issue when `OPENSSL_riscvcap=OPENSSL_riscvcap_v_zvbb_zvbc_zvkned` but not when `OPENSSL_riscvcap=rv64gcv_zvbb_zvbc_zvkned`. It seems that the provided `OPENSSL_riscvcap` variable is wrong.

Can you try with `OPENSSL_riscvcap=rv64gcv_zvbb_zvbc_zvkned`?

A summary of my tests are:
```terminal
$ export OPENSSL_riscvcap=OPENSSL_riscvcap_v_zvbb_zvbc_zvkned
$ qemu-riscv64 -cpu rv64,g=false,f=true,d=true,v=true,vlen=128,elen=64,vext_spec=v1.0,zvbb=true,zvbc=true,zvkg=true,zvknha=true,zvknhb=true,zvkned=true,zvksed=true,zvksh=true,zbkb=true,rvv_ma_all_1s=true,rvv_ta_all_1s=true -L /home/cmitrodimas/toolchains/riscv-trunk/sysroot/ ./test/evp_test ./test/recipes/30-test_evp_data/evpciph_aes_common.txt
    [...] 
    # Completed 164 tests with 40 errors and 1 skipped
    # OPENSSL_TEST_RAND_ORDER=1692031861
    not ok 1 - iteration 1
```

```terminal
$ export OPENSSL_riscvcap=rv64gcv_zvbb_zvbc_zvkned
$ qemu-riscv64 -cpu rv64,g=false,f=true,d=true,v=true,vlen=128,elen=64,vext_spec=v1.0,zvbb=true,zvbc=true,zvkg=true,zvknha=true,zvknhb=true,zvkned=true,zvksed=true,zvksh=true,zbkb=true,rvv_ma_all_1s=true,rvv_ta_all_1s=true -L /home/cmitrodimas/toolchains/riscv-trunk/sysroot/ ./test/evp_test ./test/recipes/30-test_evp_data/evpciph_aes_common.txt
    [...] 
    # Completed 164 tests with 0 errors and 1 skipped
    ok 1 - iteration 1
```

All tests with [this revision of QEMU](https://github.com/cmuellner/qemu/tree/riscv-zvk-v7).

Let me know if this helps.

phoebesv commented:
> I can reproduce your issue when `OPENSSL_riscvcap=OPENSSL_riscvcap_v_zvbb_zvbc_zvkned` but not when `OPENSSL_riscvcap=rv64gcv_zvbb_zvbc_zvkned`. It seems that the provided `OPENSSL_riscvcap` variable is wrong.
> 
> Can you try with `OPENSSL_riscvcap=rv64gcv_zvbb_zvbc_zvkned`?
> 
> A summary of my tests are:
> 
> ```
> $ export OPENSSL_riscvcap=OPENSSL_riscvcap_v_zvbb_zvbc_zvkned
> $ qemu-riscv64 -cpu rv64,g=false,f=true,d=true,v=true,vlen=128,elen=64,vext_spec=v1.0,zvbb=true,zvbc=true,zvkg=true,zvknha=true,zvknhb=true,zvkned=true,zvksed=true,zvksh=true,zbkb=true,rvv_ma_all_1s=true,rvv_ta_all_1s=true -L /home/cmitrodimas/toolchains/riscv-trunk/sysroot/ ./test/evp_test ./test/recipes/30-test_evp_data/evpciph_aes_common.txt
>     [...] 
>     # Completed 164 tests with 40 errors and 1 skipped
>     # OPENSSL_TEST_RAND_ORDER=1692031861
>     not ok 1 - iteration 1
> ```
> 
> ```
> $ export OPENSSL_riscvcap=rv64gcv_zvbb_zvbc_zvkned
> $ qemu-riscv64 -cpu rv64,g=false,f=true,d=true,v=true,vlen=128,elen=64,vext_spec=v1.0,zvbb=true,zvbc=true,zvkg=true,zvknha=true,zvknhb=true,zvkned=true,zvksed=true,zvksh=true,zbkb=true,rvv_ma_all_1s=true,rvv_ta_all_1s=true -L /home/cmitrodimas/toolchains/riscv-trunk/sysroot/ ./test/evp_test ./test/recipes/30-test_evp_data/evpciph_aes_common.txt
>     [...] 
>     # Completed 164 tests with 0 errors and 1 skipped
>     ok 1 - iteration 1
> ```
> 
> All tests with [this revision of QEMU](https://github.com/cmuellner/qemu/tree/riscv-zvk-v7).
> 
> Let me know if this helps.

@charmitro I have tried your command `OPENSSL_riscvcap=rv64gcv_zvbb_zvbc_zvkned`, it does not run to any RVV implementation, it does not seem to be correct. 
Could you please try this `OPENSSL_riscvcap=rv64gc_v_zvbb_zvbc_zvkned` instead? 
A summary of my tests are:
```
$ export OPENSSL_riscvcap=rv64gc_v_zvbb_zvbc_zvkned
$ qemu-riscv64 -cpu rv64,g=false,f=true,d=true,v=true,vlen=128,elen=64,vext_spec=v1.0,x-zvbb=true,x-zvbc=true,x-zvkg=true,x-zvknha=true,x-zvknhb=true,x-zvkned=true,x-zvksed=true,x-zvksh=true,zbkb=true,rvv_ma_all_1s=true,rvv_ta_all_1s=true ./test/evp_test ./test/recipes/30-test_evp_data/evpciph_aes_common.txt
[...]
    # Completed 164 tests with 40 errors and 1 skipped
    # OPENSSL_TEST_RAND_ORDER=1692036246
    not ok 1 - iteration 1
```

ardbiesheuvel commented:
I have submitted the missing ICLA for ardb@google.com to legal@ today.

cmuellner commented:
> I think the root cause of the failure of the test is the incorrect mask policy in `ghash-riscv64-zvbb-zvbc.pl`. It should be set as `mask-undisturbed`. Tail/mask-agnostic setting should be aware, since some CPUs will force it to a default value. You can test if the tail or mask policy is incorrect by adding `rvv_ma_all_1s=true and rvv_ta_all_1s=true` to the qemu argument. This is helpful for identifying potential issues with tails and masks. After running your qemu with the additional rvv_ma_all_1s=true,rvv_ta_all_1s=true parameters, I was able to reproduce the problem.

@phoebesv, I went through this in detail. You are absolutely right. Thank you for analyzing this and reporting the relevant details.

Just to provide the relevant context to others:
The tail/mask behavior for RISC-V vector can be controlled via the `vta` (tail-elements) and `vma` (masked elements) bits, which configures the behavior either to `undisturbed` or `agnostic`. If configured to `undisturbed`, then inactive elements will be left unchanged. When configured to `agnositc` the implementation is allowed to either leave the element unchanged or to overwrite them with `1`s.

The default behavior of QEMU for `agnostic` is to keep the original values. This can be changed with the mentioned CPU parameters (`rvv_ma_all_1s=true` and `rvv_ta_all_1s=true`).

I've updated the patchset to use `unused` for `vta` and `vma`, because that's the behavior that was expected when writing the code. Tests are currently running and I will push them into the PR once the tests are completed.

cmuellner commented:
This PR has been updated:

* rebased on upstream/master
* changed tail and masked element behaviour from agnostic to undisturbed in all `vsetivli` instructions

The PR has been tested with QEMU (`rvv_ma_all_1s=true,rvv_ta_all_1s=true`) in the following combinations:
* qemu-cpu-zvbb-zvbc + openssl-cap-zvbb-zvbc
* qemu-cpu-zvkg + openssl-cap-zvkg
* qemu-cpu-zvbb-zvkg + openssl-cap-zvbb-zvkg
* qemu-cpu-zvkned + openssl-cap-zvkned
* qemu-cpu-zvbb-zvknha + openssl-cap-zvbb-zvknha
* qemu-cpu-zvbb-zvknhb + openssl-cap-zvbb-zvknhb
* qemu-cpu-zvbb-zvksed + openssl-cap-zvbb-zvksed
* qemu-cpu-zvbb-zvksh + openssl-cap-zvbb-zvksh
* qemu-cpu-zvk + openssl-cap-zvk

To verify that the optimized routines are used, the following combinations have been used (must fail tests):
* must fail: qemu-cpu-base + openssl-cap-zvbb-zvbc
* must fail:  qemu-cpu-base + openssl-cap-zvkg
* must fail:  qemu-cpu-base + openssl-cap-zvbb-zvkg
* must fail:  qemu-cpu-base + openssl-cap-zvkned
* must fail:  qemu-cpu-base + openssl-cap-zvknha
* must fail:  qemu-cpu-base + openssl-cap-zvbb-zvknha
* must fail:  qemu-cpu-base + openssl-cap-zvbb-zvknhb
* must fail:  qemu-cpu-base + openssl-cap-zvbb-zvksed
* must fail:  qemu-cpu-base + openssl-cap-zvbb-zvksh

JerryShih reviewed (changes requested):
I haven't understood the math of ghash using `zvbc`. So, I can't comment that parts.
The ghash with `zvkg` is much simpler. I have checked that parts.
I also check the remaining aes, sha, sm3, sm4 functions.

paulidale commented:
> I have submitted the missing ICLA for [ardb@google.com](mailto:ardb@google.com) to legal@ today.

I didn't see it come in.  Could you try resending to osf-contact@openssl.org?

ardbiesheuvel commented:
> > I have submitted the missing ICLA for [ardb@google.com](mailto:ardb@google.com) to legal@ today.
> 
> I didn't see it come in. Could you try resending to [osf-contact@openssl.org](mailto:osf-contact@openssl.org)?

Done.

phoebesv commented on crypto/sha/sha_riscv.c:
```diff
@@ -0,0 +1,28 @@
+
+void sha256_block_data_order(SHA256_CTX *ctx, const void *in, size_t num)
+{
+    if (RISCV_HAS_ZVBB_AND_ZVKNHA() && riscv_vlen() >= 128) {
```
=>   if ((RISCV_HAS_ZVBB_AND_ZVKNHA() || RISCV_HAS_ZVBB_AND_ZVKNHB()) && riscv_vlen() >= 128) 
Zvknhb also supports SHA-256.

phoebesv commented on crypto/aes/asm/aes-riscv64-zvkned.pl:
```diff
@@ -0,0 +1,637 @@
+
+    @{[vle32_v $v1, ($INP)]}
+
+    @{[vaesz_vs $v1, $v20]}    # with round key w[43,47]
```
\# with round key w[43,47] => \# with round key w[40,43]

JerryShih commented:
We already have the merged pr(https://github.com/openssl/openssl/pull/21923) containing this pr work. I think we could close this pr now.
